File: Acwing/Basic Algorithm/Basic/802. interval_and/Main.cpp
#include<iostream>
#include<algorithm>
#include<vector>
using namespace std;

typedef pair<int,int> PII;

const int N = 300010;
int a[N] , s[N];
vector<int> alls;
vector<PII> add , query;

int find(int x){
    int l = 0 , r = alls.size() - 1;
    while (l < r){
        int mid = (l + r) >> 1;
        if(alls[mid] >= x) r = mid;
        else l = mid + 1;
    }
    return r + 1;
}

int main(){
    int n , m;
    cin >> n >> m;
    int x , c;
    for(int i = 0 ; i < n ; i++){
        cin >> x >> c;
        alls.push_back(x);
        add.push_back({x , c});
    }
    int l , r;
    for(int i = 0 ; i < m ; i++){
        cin >> l >> r;
        alls.push_back(l);
        alls.push_back(r);
        query.push_back({l , r});
    }
    
    sort(alls.begin(),alls.end());
    alls.erase(unique(alls.begin(),alls.end()),alls.end());
    
    for( auto item : add){
        x = find(item.first);
        a[x] += item.second;
    }
    
    for(int i = 1 ; i <= alls.size(); i++) s[i] = s[i-1] + a[i];
    
    for( auto item : query){
        l = find(item.first) , r = find(item.second);
        cout << s[r] - s[l - 1] << endl;
    }
    return 0;
    
}

File: Acwing/Basic Algorithm/Basic/802. interval_and/Main.py
def find(array, x : int):
    """返回x在array中的下标"""
    l = 0
    r = len(array) - 1
    while l < r:
        mid = (l + r + 1) >> 1
        if array[mid] <= x:
            l = mid
        else:
            r = mid - 1
    return l + 1

nm = [int(i) for i in input().split()]
n, m = nm[0], nm[1]
adds = [] 
query = []
alls = [] # 待离散化的数据
a = [0 for i in range(3 * 100010)] # 离散化后的结果
for i in range(n):
    add = [int(i) for i in input().split()]
    adds.append(add)
    alls.append(add[0])
# print(adds)
for i in range(m):
    queryTemp = [int(i) for i in input().split()]
    query.append(queryTemp)
    alls.append(queryTemp[0])
    alls.append(queryTemp[1])
# 去重+排序
alls = list(set(alls))
alls = sorted(alls)
#处理插入
for item in adds:
    x = find(alls, item[0])
    a[x] += item[1]

# 计算前缀和
s = [0 for i in range(3 * 100010)]
for i in range(1, len(a)):
    s[i] = s[i - 1] + a[i]
# 处理查询
for item in query:
    l : int = find(alls, item[0])
    r : int = find(alls, item[1])
    print(s[r] - s[l - 1])
